uniffi_core/src/ffi/foreigncallbacks.rs

```diff
@@ -142,162 +37,67 @@ pub type ForeignCallback = unsafe extern "C" fn(
-    /// The error value is serialized to `buf_ptr`.
-    Error = 1,
-    /// Unexpected error.
-    /// An error message string is serialized to `buf_ptr`.
-    UnexpectedError = 2,
-}
-
-impl TryFrom<i32> for CallbackResult {
-    // On errors we return the unconverted value
-    type Error = i32;
-
-    fn try_from(value: i32) -> Result<Self, i32> {
-        match value {
-            0 => Ok(Self::Success),
-            1 => Ok(Self::Error),
-            2 => Ok(Self::UnexpectedError),
-            n => Err(n)
-        }
-    }
-}
-
-// Overly-paranoid sanity checking to ensure that these types are
-// convertible between each-other. `transmute` actually should check this for
-// us too, but this helps document the invariants we rely on in this code.
-//
-// Note that these are guaranteed by
-// https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html
-// and thus this is a little paranoid.
-static_assertions::assert_eq_size!(usize, ForeignCallback);
-static_assertions::assert_eq_size!(usize, Option<ForeignCallback>);
-
-/// Struct to hold a foreign callback.
-pub struct ForeignCallbackInternals {
-    callback_ptr: AtomicUsize,
-}
-
-const EMPTY_PTR: usize = 0;
-
-impl ForeignCallbackInternals {
-    pub const fn new() -> Self {
-        ForeignCallbackInternals {
-            callback_ptr: AtomicUsize::new(EMPTY_PTR),
-        }
-    }
-
-    pub fn set_callback(&self, callback: ForeignCallback) {
-        let as_usize = callback as usize;
-        let old_ptr = self.callback_ptr.compare_exchange(
-            EMPTY_PTR,
-            as_usize,
-            Ordering::SeqCst,
-            Ordering::SeqCst,
-        );
-        match old_ptr {
-            // We get the previous value back. If this is anything except EMPTY_PTR,
-            // then this has been set before we get here.
-            Ok(EMPTY_PTR) => (),
-            _ =>
-            // This is an internal bug, the other side of the FFI should ensure
-            // it sets this only once.
-            {
-                panic!("Bug: call set_callback multiple times. This is likely a uniffi bug")
+/// Callback to schedule a Rust call with a `ForeignExecutor`. The bindings code registers exactly
+/// one of these with the Rust code.
+///
+/// Delay is an approximate amount of ms to wait before scheduling the call.  Delay is usually 0,
+/// which means schedule sometime soon.
+///
+/// As a special case, when Rust drops the foreign executor, with `task=null`.  The foreign
+/// bindings should release the reference to the executor that was reserved for Rust.
+///
+/// This callback can be invoked from any thread, including threads created by Rust.
+///
+/// The callback should return one of the `ForeignExecutorCallbackResult` values.
+pub type ForeignExecutorCallback = extern "C" fn(
+    executor: ForeignExecutorHandle,
+    delay: u32,
+    task: Option<RustTaskCallback>,
+    task_data: *const (),
+) -> i8;
```

Comment:
I went back and forth on this one, but since the callback is implemented in a foreign language I think it should be the raw value.  If the foreign code sends back an invalid value, we can panic with a somewhat nice message rather than enter UB-land.